### DISCUSSION TITLE
Set CPU usage for BrainVISA on HPC

### DISCUSSION THREAD

**Discussion by wsnyder4** [brainvisa/brainvisa.github.io#70] [Q&A] [answered]
Dear BrainVISA experts and developers,

I am looking to run the BrainVISA/morphologist pipeline on the UK Biobank, and have successfully run the pipeline on test subjects using a singularity image transferred to / installed on a high performance computing cluster. However, these test runs currently will use up the maximum number of CPUs available, and I would like to limit to one or two CPUs instead. 

Due to some constraints on this HPC, I may be unable to use soma-workflow tools to manage jobs. Without soma-workflow, how may I be able to limit the number of CPUs BrainVISA uses within singularity? I do not have admin access either and therefore cannot include the number of CPUs as an option in a singularity command.

Thank you in advance for the assistance!
Best,
Will Snyder

**Comment by denisri** [comment]
Hi,

> Due to some constraints on this HPC, I may be unable to use soma-workflow tools to manage jobs.

You mean that you are not allow to run a server process on the front-end machine of the cluster (fr soma-workflow servers), or are limited to a single job, or your HCP jobs manager is not supported by Soma-Workflow ?
Or you just don't have set a jobs limit is SWF which submits more jobs than you are allowed ? In the latter cases, there are options to limit that (in the GUI and in the config file).

Then how are you running the pipeline ? By submiting "by hand" (or by a script) a job for each Morphologist pipeline ?
If you are running the pipeline using the axon-runprocess tool, and don't specify the `--swf` option, then the pipeline should run sequentially, using a single CPU, except for processings that are internally parallelized: a few steps use internal parallelism (threading), like the "cortical folds graph" and "sulci recognition using CNN" steps. They both have an option to allow/disable threading: `allow_multithreading` (which is ON by default, and there is no fine-grain option to use a given number of CPUs). The normalization step using SPM is also using parallelism I think, but I don't know how to disable it.

Denis

**Reply by wsnyder4** [reply]
Turning off multithreading worked well for my purposes, thank you so much!
My HPC prefers that I run all code in temporary directories assigned upon each job submission (one job per subject pipeline) and so I have been using this less-conventional approach to allow customization of where in's and out's go.

**Comment by ylep** [comment, marked as answer]
Maybe off-topic, but you could use the `taskset` command as a very general way of restricting a process and its children to run on a subset of a system's CPUs ([see the taskset(1) manual](https://www.man7.org/linux/man-pages/man1/taskset.1.html)).
